src/main/java/graphql/analysis/MaxQueryComplexityInstrumentation.java

```diff
@@ -76,41 +81,52 @@ public MaxQueryComplexityInstrumentation(int maxComplexity, FieldComplexityCalcu
         this.maxQueryComplexityExceededFunction = maxQueryComplexityExceededFunction;
     }
 
+    @Override
+    public InstrumentationState createState(InstrumentationCreateStateParameters parameters) {
+        return new State();
+    }
+
     @Override
     public InstrumentationContext<List<ValidationError>> beginValidation(InstrumentationValidationParameters parameters) {
-        return whenCompleted((errors, throwable) -> {
-            if ((errors != null && errors.size() > 0) || throwable != null) {
-                return;
-            }
-            QueryTraverser queryTraverser = newQueryTraverser(parameters);
-
-            Map<QueryVisitorFieldEnvironment, Integer> valuesByParent = new LinkedHashMap<>();
-            queryTraverser.visitPostOrder(new QueryVisitorStub() {
-                @Override
-                public void visitField(QueryVisitorFieldEnvironment env) {
-                    int childsComplexity = valuesByParent.getOrDefault(env, 0);
-                    int value = calculateComplexity(env, childsComplexity);
-
-                    valuesByParent.compute(env.getParentEnvironment(), (key, oldValue) ->
-                            ofNullable(oldValue).orElse(0) + value
-                    );
-                }
-            });
-            int totalComplexity = valuesByParent.getOrDefault(null, 0);
-            if (log.isDebugEnabled()) {
-                log.debug("Query complexity: {}", totalComplexity);
-            }
-            if (totalComplexity > maxComplexity) {
-                QueryComplexityInfo queryComplexityInfo = QueryComplexityInfo.newQueryComplexityInfo()
-                        .complexity(totalComplexity)
-                        .instrumentationValidationParameters(parameters)
-                        .build();
-                boolean throwAbortException = maxQueryComplexityExceededFunction.apply(queryComplexityInfo);
-                if (throwAbortException) {
-                    throw mkAbortException(totalComplexity, maxComplexity);
-                }
+        State state = parameters.getInstrumentationState();
+        // for API backwards compatibility reasons we capture the validation parameters, so we can put them into QueryComplexityInfo
+        state.instrumentationValidationParameters = parameters;
```

Comment:
this could be read in another Thread I think. Therefore `instrumentationValidationParameters` would need to synchronized or Atomic I think

Reply:
fixed - in reality it would not be a problem since the 2 methods are called within the same thread - but hey - I fixed it anyway